### PR TITLE
Task3

### PR DESCRIPTION
**Pull request by fujoshik**

### CODE REVIEW

File: Task 2/Task2.3/Trainings/Training.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                }
+            }
+            return true;
+        }
+
+        public Training Clone()
+        {
+            var cloned = new Training(Description);
+
+            foreach (var training in _trainings)
```
Comment (IgorKheidorov) [inline Task 2/Task2.3/Trainings/Training.cs]: if-else and casting are not needed, just cloned.Add((BaseTraining)training.Clone());

File: Task 2/Task2.3/Trainings/Training.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+﻿namespace Task2._3.Trainings
+{
+    public class Training : TrainingEntity
+    {
+        private int _trainingsCount;
```
Comment (IgorKheidorov) [inline Task 2/Task2.3/Trainings/Training.cs]: _trainings not perfect name, better _lessons or _trainingEntitys

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Please fix comments. More, try not mix files from different tasks in one PR. In this PR for task 3 I see files for the next task. To avoid this situation start new branch for a task not from the previous branch, but from master

File: Task 3/Task3/Queue.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        private int _size;
+
+        public Queue(int size = 10)
+        {
+            _size = size;
+            _queue = new T[size];
+        }
+
+        public T Dequeue()
+        {
```
Comment (IgorKheidorov) [inline Task 3/Task3/Queue.cs:17]: Good solution idea, but not fully implemented. Think about if we make 10  Enqueue, then 10 Dequeue. Really stack is empty, but in you implementation _head = _tail = _size. and new Enque is not more possible. Please fix ))

File: Task 3/Task3/QueueExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿namespace Task3
+{
+    public static class QueueExtensions
+    {
+        public static IQueue<T> Tail<T>(this IQueue<T> queue, int length = 10)
```
Comment (IgorKheidorov) [inline Task 3/Task3/QueueExtensions.cs]: In this implementation you change the original queue (as result it is empty). But this method has to create new queue not changing the original one

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Please fix comments

File: Task 2/Task2.3/Trainings/Training.cs (hunk 1 above)
Comment (fujoshik) [reply Task 2/Task2.3/Trainings/Training.cs]: Fixed

File: Task 2/Task2.3/Trainings/Training.cs (hunk 2 above)
Comment (fujoshik) [reply Task 2/Task2.3/Trainings/Training.cs]: Fixed

File: Task 3/Task3/Queue.cs (hunk 3 above)
Comment (fujoshik) [reply Task 3/Task3/Queue.cs:17]: Fixed

File: Task 3/Task3/QueueExtensions.cs (hunk 4 above)
Comment (fujoshik) [reply Task 3/Task3/QueueExtensions.cs]: Fixed

File: Task 3/Task3/IQueue.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Task3
+{
```
Comment (IgorKheidorov) [inline Task 3/Task3/IQueue.cs:2]: Restrict o be used for value types only

File: Task 3/Task3/Queue.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+            if (object.Equals(element, default(T)))
+            {
+                throw new InvalidOperationException("No elements in the queue!");
+            }
+            _head++;
+
+            return element;
+        }
+
```
Comment (IgorKheidorov) [inline Task 3/Task3/Queue.cs:32]: The issue is the same- I can Enqueue only 10 times, not depending on the position of head

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Please fix comments

File: Task 3/Task3/Queue.cs (hunk 6 above)
Comment (fujoshik) [reply Task 3/Task3/Queue.cs:32]: I checked it in Program.cs with queue2 (with size of 5 elements) and I was able to Enqueue 5 times, then Dequeue 5 times and then again Enqueue more, so I think it should work

File: Task 3/Task3/IQueue.cs (hunk 5 above)
Comment (fujoshik) [reply Task 3/Task3/IQueue.cs:2]: Fixed

**Review by IgorKheidorov — APPROVED** [review]
Cool

**Pull request merged**